### PR TITLE
Add falcon40b e2e perf tests to ci

### PR DESCRIPTION
**Pull request by johanna-rock-tt**
t3k perf pipeline [passing]: https://github.com/tenstorrent/tt-metal/actions/runs/9467522099

### CODE REVIEW

File: models/demos/t3000/falcon40b/tests/test_perf_falcon.py [hunk 1]
```diff
@@ -199,13 +208,26 @@ def run_test_FalconCausalLM_end_to_end(
...
+        )
+        attention_mask_memconfig = model_config["ATTN_MASK_MEMCFG"]
+        num_max_tokens = nearest_32(kv_cache_len + 1)
+        if attention_mask_memconfig.is_sharded():
+            attn_mask_shard_shape = attention_mask_memconfig.shard_spec.shape
+            attn_mask_shard_shape[-1] = num_max_tokens
+            attention_mask_memconfig.shard_spec.shape = attn_mask_shard_shape
         tt_attention_mask = [
             tt_attention_mask_host[i].to(devices[i], attention_mask_memconfig) for i in range(len(devices))
         ]
```
Comment (skhorasganiTT) [inline models/demos/t3000/falcon40b/tests/test_perf_falcon.py:222]: This is already done in model_preprocessing and the mask is already on device right?

File: models/demos/t3000/falcon40b/tests/test_perf_falcon.py [hunk 2]
```diff
@@ -247,13 +272,38 @@ def run_test_FalconCausalLM_end_to_end(
...
+        )
+        attention_mask_memconfig = model_config["ATTN_MASK_MEMCFG"]
+        num_max_tokens = nearest_32(kv_cache_len + 1)
+        if attention_mask_memconfig.is_sharded():
+            attn_mask_shard_shape = attention_mask_memconfig.shard_spec.shape
+            attn_mask_shard_shape[-1] = num_max_tokens
+            attention_mask_memconfig.shard_spec.shape = attn_mask_shard_shape
+        tt_attention_mask = [
+            tt_attention_mask_host[i].to(devices[i], attention_mask_memconfig) for i in range(len(devices))
+        ]
```
Comment (skhorasganiTT) [inline models/demos/t3000/falcon40b/tests/test_perf_falcon.py:296]: Same comment as above

File: models/demos/t3000/falcon40b/tests/test_perf_falcon.py [hunk 3]
```diff
@@ -239,6 +261,9 @@ def run_test_FalconCausalLM_end_to_end(
                 layer_past_len=kv_cache_len,
                 use_cache=use_cache,
             )
+            tt_inputs, tt_attention_mask_host = tt_FalconCausalLM.model_preprocessing(
+                llm_mode, model_input, kv_cache_len, num_input_tokens=kv_len
+            )
             tt_attention_mask = [
                 tt_attention_mask_host[i].to(devices[i], attention_mask_memconfig) for i in range(len(devices))
             ]
```
Comment (skhorasganiTT) [inline models/demos/t3000/falcon40b/tests/test_perf_falcon.py:269]: Same as above, mask returned by model_preprocessing is already on device

File: models/demos/t3000/falcon40b/tests/test_perf_falcon.py [hunk 4]
```diff
@@ -337,20 +387,19 @@ def run_test_FalconCausalLM_end_to_end(
...
     ("tiiuae/falcon-40b-instruct",),
     ids=["falcon_40b"],
 )
-@pytest.mark.parametrize("model_config_str", ("BFLOAT8_B-SHARDED", "BFLOAT8_B-DRAM", "BFLOAT16-DRAM"))
+@pytest.mark.parametrize(
+    "model_config_str",
+    (
+        "BFLOAT8_B-SHARDED",
+        "BFLOAT8_B-DRAM",
+        # "BFLOAT16-DRAM"
```
Comment (skhorasganiTT) [inline models/demos/t3000/falcon40b/tests/test_perf_falcon.py:400]: Should BFLOAT16-DRAM be kept here since it's used for prefill in the demo?

File: tests/scripts/t3000/run_t3000_model_perf_tests.sh [hunk 5]
```diff
@@ -44,6 +44,20 @@ run_t3000_llama2_70b_tests() {
   echo "LOG_METAL: run_t3000_llama2_70b_tests $duration seconds to complete"
 }
 
+run_t3000_falcon40b_tests() {
+  # Record the start time
+  start_time=$(date +%s)
+
+  echo "LOG_METAL: Running run_t3000_falcon40b_tests"
+
+  env WH_ARCH_YAML=wormhole_b0_80_arch_eth_dispatch.yaml pytest models/demos/t3000/falcon40b/tests/test_perf_falcon.py -m "model_perf_t3000"
```
Comment (skhorasganiTT) [inline tests/scripts/t3000/run_t3000_model_perf_tests.sh:53]: I don' think Falcon40 was run in the posted run, is "model_perf_t3000" incorrect here?

File: models/demos/t3000/falcon40b/tests/test_perf_falcon.py (hunk 1 above)
Comment (johanna-rock-tt) [reply models/demos/t3000/falcon40b/tests/test_perf_falcon.py]: Yeah you're right! Thanks for pointing that out. I removed all the attention mask processing in the perf test file now, it was all redundant.

File: models/demos/t3000/falcon40b/tests/test_perf_falcon.py (hunk 4 above)
Comment (johanna-rock-tt) [reply models/demos/t3000/falcon40b/tests/test_perf_falcon.py]: I enabled `BFLOAT16-DRAM` and added the perf targets for it.

I also changed the demo to use `BFLOAT8_B-DRAM` config for prefill since this is our main supported version atm. PCC is >0.99 since we changed the residual path to be fp16 in that version. I also ran the demo with some example inputs and output looks good.

**Review by djordje-tt — APPROVED** [review]

**Review by skhorasganiTT — APPROVED** [review]

**Review by ttmchiou — APPROVED** [review]

**Review by s-jovic — APPROVED** [review]

**Comment by johanna-rock-tt** [comment]
@s-jovic The targets should be fine. Yeah you're right it's the output that's 0.99 and kv cache targets are a bit lower. The point was that bfp8 is very similar to fp16 in terms of pcc now, so I think we can safely use it for the demo.
